modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java

```diff
@@ -184,60 +233,113 @@ private CompletableFuture<?> onIndexRemoved(RemoveIndexEventParameters parameter
-                    return nullCompletedFuture();
-                }
+                if (parameters.groupId() instanceof ZonePartitionId) {
+                    assert enabledColocation() : "Primary replica ID must be of type ZonePartitionId";
 
-                return getMvTableStorageFuture(parameters.causalityToken(), primaryReplicaId)
-                        .thenCompose(mvTableStorage -> awaitPrimaryReplica(primaryReplicaId, parameters.startTime())
-                                .thenAccept(replicaMeta -> tryScheduleBuildIndexesForNewPrimaryReplica(
-                                        catalog.version(),
-                                        primaryReplicaId,
-                                        mvTableStorage,
-                                        replicaMeta
-                                ))
-                        );
+                    ZonePartitionId primaryReplicaId = (ZonePartitionId) parameters.groupId();
+
+                    CatalogZoneDescriptor zoneDescriptor = catalog.zone(primaryReplicaId.zoneId());
+                    // TODO: IGNITE-22656 It is necessary not to generate an event for a destroyed zone by LWM
+                    if (zoneDescriptor == null) {
+                        return nullCompletedFuture();
+                    }
+
+                    var indexFutures = new ArrayList<CompletableFuture<?>>();
+                    for (CatalogTableDescriptor tableDescriptor : catalog.tables()) {
+                        // 1. Perhaps, it makes sense to get primary replica future first and then get table storage future,
+                        // because, it will be the same for all tables in the zone for the given partition.
+                        // 2. Is it possible to filter out tables in efficient way
+                        // that do not have indices to avoid creating unnecessary futures?
+                        // It looks like catalog.indexes(tableDescriptor.id()).isEmpty() is good enough.
+                        // However, what about PK index?
+                        CompletableFuture<?> future =
+                                getMvTableStorageFuture(parameters.causalityToken(), tableDescriptor.id())
+                                        .thenCompose(mvTableStorage -> awaitPrimaryReplica(primaryReplicaId, parameters.startTime())
+                                                .thenAccept(replicaMeta -> tryScheduleBuildIndexesForNewPrimaryReplica(
+                                                        catalog.version(),
+                                                        tableDescriptor,
+                                                        primaryReplicaId,
+                                                        mvTableStorage,
+                                                        replicaMeta)));
+
+                        indexFutures.add(future);
+                    }
+
+                    return allOf(indexFutures.toArray(CompletableFuture[]::new));
```

Comment:
U can use `CompletableFutures#allOf`

Reply:
Please note: there is `CompletableFuture#allOf()` from the standard library which you use here, it accepts an array; and there is `CompletableFutures#allOf()` defined in our code (mind the 's' in which the class name ends) which accepts a collection

Reply:
Got it! Thank you, guys!